test/hotspot/jtreg/compiler/gcbarriers/TestZGCBarrierElision.java

```diff
@@ -0,0 +1,359 @@
+        if (value % 2 == 0) {
+            o.field1 = i;
+        }
+        o.field1 = i;
+    }
+
+    @Test
+    @IR(counts = { IRNode.Z_STORE_P_WITH_BARRIER_FLAG, REMAINING, "1" }, phase = CompilePhase.FINAL_CODE)
+    @IR(counts = { IRNode.Z_STORE_P_WITH_BARRIER_FLAG, ELIDED, "1" }, phase = CompilePhase.FINAL_CODE)
+    static void testStoreThenStoreInLoop(Outer o, Inner i) {
+        o.field1 = i;
+        for (int j = 0; j < 100; j++) {
+            o.field1 = i;
+        }
+    }
+
+    @Test
+    @IR(counts = { IRNode.Z_STORE_P_WITH_BARRIER_FLAG, REMAINING, "2" }, phase = CompilePhase.FINAL_CODE)
+    static void testStoreThenCallThenStore(Outer o, Inner i) {
+        o.field1 = i;
+        nonInlinedMethod();
+        o.field1 = i;
+    }
+
+    @Run(test = {"testStoreThenConditionalStore",
+                 "testConditionalStoreThenStore",
+                 "testStoreThenStoreInLoop",
+                 "testStoreThenCallThenStore"})
+    void runControlFlowTests() {
+        testStoreThenConditionalStore(outer, inner, ThreadLocalRandom.current().nextInt(0, 100));
+        testConditionalStoreThenStore(outer, inner, ThreadLocalRandom.current().nextInt(0, 100));
+        testStoreThenStoreInLoop(outer, inner);
+        testStoreThenCallThenStore(outer, inner);
+    }
+
+    @Test
+    @IR(counts = { IRNode.Z_GET_AND_SET_P_WITH_BARRIER_FLAG, REMAINING, "1" }, phase = CompilePhase.FINAL_CODE)
+    static void testAllocateThenAtomic(Inner i) {
+        Outer o = new Outer();
+        blackhole(o);
+        field1VarHandle.getAndSet​(o, i);
+    }
+
+    @Test
+    @IR(counts = { IRNode.Z_LOAD_P_WITH_BARRIER_FLAG, REMAINING, "1" }, phase = CompilePhase.FINAL_CODE)
+    @IR(counts = { IRNode.Z_GET_AND_SET_P_WITH_BARRIER_FLAG, REMAINING, "1" }, phase = CompilePhase.FINAL_CODE)
+    static void testLoadThenAtomic(Outer o, Inner i) {
+        blackhole(o.field1);
+        field1VarHandle.getAndSet​(o, i);
+    }
+
+    @Test
+    @IR(counts = { IRNode.Z_STORE_P_WITH_BARRIER_FLAG, REMAINING, "1" }, phase = CompilePhase.FINAL_CODE)
+    // The atomic access barrier should be elided, but is not.
+    static void testStoreThenAtomic(Outer o, Inner i) {
+        o.field1 = i;
+        field1VarHandle.getAndSet​(o, i);
+    }
+
+    @Test
+    @IR(counts = { IRNode.Z_GET_AND_SET_P_WITH_BARRIER_FLAG, REMAINING, "1" }, phase = CompilePhase.FINAL_CODE)
+    // The load barrier should be elided, but is not.
+    static void testAtomicThenLoad(Outer o, Inner i) {
+        field1VarHandle.getAndSet​(o, i);
+        blackhole(o.field1);
+    }
+
+    @Test
+    @IR(counts = { IRNode.Z_GET_AND_SET_P_WITH_BARRIER_FLAG, REMAINING, "1" }, phase = CompilePhase.FINAL_CODE)
+    // The store barrier should be elided, but is not.
+    static void testAtomicThenStore(Outer o, Inner i) {
+        field1VarHandle.getAndSet​(o, i);
+        o.field1 = i;
+    }
+
+    @Test
+    // The second atomic access barrier should be elided, but is not.
```

Comment:
Is it worth explicitly documenting the behaviour and the wanted/expected behaviour in annotations?
E.g.
```java
    @Test
    @IR(counts = { IRNode.Z_GET_AND_SET_P_WITH_BARRIER_FLAG, REMAINING, "2" }, phase = CompilePhase.FINAL_CODE)
    @IR(counts = { IRNode.Z_GET_AND_SET_P_WITH_BARRIER_FLAG, ELIDED, "0" }, phase = CompilePhase.FINAL_CODE)
    // The second atomic access barrier should be elided, but is not.
    // @IR(counts = { IRNode.Z_GET_AND_SET_P_WITH_BARRIER_FLAG, REMAINING, "1" }, phase = CompilePhase.FINAL_CODE)
    // @IR(counts = { IRNode.Z_GET_AND_SET_P_WITH_BARRIER_FLAG, ELIDED, "1" }, phase = CompilePhase.FINAL_CODE)
    static void testAtomicThenAtomic(Outer o, Inner i) {
        field1VarHandle.getAndSet​(o, i);
        field1VarHandle.getAndSet​(o, i);
    }
```

Same goes for all the other test cases.

Reply:
IR checks in compiler tests are commonly interpreted as documentation for expected behavior only. I fear adding IR checks for the current (unexpected) behavior would risk misleading inattentive readers, so I would rather leave it as-is if you don't mind.